### ISSUE TITLE
trying to implement ReflectionPadding2D

### ISSUE THREAD

**Issue by iperov** [plaidml/plaidml#273]
finally I implemented it in one op for B,H,W,C
```
class ReflectionPadding2D(PMLTile.Operation):
    def __init__(self, input, h_pad, w_pad):
        if K.image_data_format() == 'channels_last':
            if input.shape.ndims == 4:
                H, W = input.shape.dims[1:3]
                if (type(H) == int and h_pad >= H) or \
                   (type(W) == int and w_pad >= W):
                    raise ValueError("Paddings must be less than dimensions.")
                c = """ function (I[B, H, W, C] ) -> (O) {{
                        WE = W + {w_pad}*2;
                        HE = H + {h_pad}*2;
                    """.format(h_pad=h_pad, w_pad=w_pad)
                if w_pad > 0:
                    c += """
                        LEFT_PAD [b, h, w , c : B, H, WE, C ] = =(I[b, h, {w_pad}-w,            c]), w < {w_pad} ;
                        HCENTER  [b, h, w , c : B, H, WE, C ] = =(I[b, h, w-{w_pad},            c]), w < W+{w_pad}-1 ;
                        RIGHT_PAD[b, h, w , c : B, H, WE, C ] = =(I[b, h, 2*W - (w-{w_pad}) -2, c]);
                        LCR = LEFT_PAD+HCENTER+RIGHT_PAD;
                    """.format(h_pad=h_pad, w_pad=w_pad)
                else:
                    c += "LCR = I;"
                if h_pad > 0:
                    c += """
                        TOP_PAD   [b, h, w , c : B, HE, WE, C ] = =(LCR[b, {h_pad}-h,            w, c]), h < {h_pad};
                        VCENTER   [b, h, w , c : B, HE, WE, C ] = =(LCR[b, h-{h_pad},            w, c]), h < H+{h_pad}-1 ;
                        BOTTOM_PAD[b, h, w , c : B, HE, WE, C ] = =(LCR[b, 2*H - (h-{h_pad}) -2, w, c]);
                        TVB = TOP_PAD+VCENTER+BOTTOM_PAD;
                    """.format(h_pad=h_pad, w_pad=w_pad)
                else:
                    c += "TVB = LCR;"
                c += "O = TVB; }"
                inp_dims = input.shape.dims
                out_dims = (inp_dims[0], inp_dims[1]+h_pad*2, inp_dims[2]+w_pad*2, inp_dims[3])
            else:
                raise NotImplemented
        else:
            raise NotImplemented
        super(ReflectionPadding2D, self).__init__(c, [('I', input) ],
                [('O', PMLTile.Shape(input.shape.dtype, out_dims ) )])
```

also I implemented it via slice and concat
but I suppose it will consume more VRAM for this? or am I wrong??
```
class ReflectionPadding2D():
    def __init__(self, h_pad, w_pad):
        self.h_pad, self.w_pad = h_pad, w_pad
    def __call__(self, inp):
        h_pad, w_pad = self.h_pad, self.w_pad
        if K.image_data_format() == 'channels_last':
            if inp.shape.ndims == 4:
                w = K.concatenate ([ inp[:,:,w_pad:0:-1,:],
                                     inp,
                                     inp[:,:,-2:-w_pad-2:-1,:] ], axis=2 )
                h = K.concatenate ([ w[:,h_pad:0:-1,:,:],
                                     w,
                                     w[:,-2:-h_pad-2:-1,:,:] ], axis=1 )
                return h
            else:
                raise NotImplemented
        else:
            raise NotImplemented
```

**Comment by iperov** [comment]
I can implement it via slice and concat
but I suppose it will consume more VRAM for this? or am I wrong??
```
class ReflectionPadding2D():
    def __init__(self, h_pad, w_pad):
        self.h_pad, self.w_pad = h_pad, w_pad
    def __call__(self, inp):
        h_pad, w_pad = self.h_pad, self.w_pad
        if K.image_data_format() == 'channels_last':
            if inp.shape.ndims == 4:
                w = K.concatenate ([ inp[:,:,w_pad:0:-1,:],
                                     inp,
                                     inp[:,:,-2:-w_pad-2:-1,:] ], axis=2 )
                h = K.concatenate ([ w[:,h_pad:0:-1,:,:],
                                     w,
                                     w[:,-2:-h_pad-2:-1,:,:] ], axis=1 )
                return h
            else:
                raise NotImplemented
        else:
            raise NotImplemented
```

**Comment by iperov** [comment]
are both version the same in performance and consumption of VRAM ?

**Comment by iperov** [comment]
tested both versions
one op version vs slice-concat version : much less ops to analyze, but ~3% slower

**Comment by SleepProgger** [comment]
In case someone needs it: i added support for left/right and top/down padding to be more compatible to tensorflow to iperovs version.
If reflected padding would be added to plaidml i think a version working for all dimension like `tf.pad` would be better though.

```
class ReflectionPadding2D(plaidml.tile.Operation):
    def __init__(self, input_, h_pad, w_pad):
        if K.image_data_format() == 'channels_last':
            if input_.shape.ndims == 4:
                H, W = input_.shape.dims[1:3]
                ws_pad = w_pad[0] + w_pad[1]
                hs_pad = h_pad[0] + h_pad[1]
                if (type(H) == int and hs_pad >= H) or \
                   (type(W) == int and ws_pad >= W):
                    raise ValueError("Paddings must be less than dimensions.")
                c = """ function (I[B, H, W, C] ) -> (O) {{
                        WE = W + {w_pad};
                        HE = H + {h_pad};
                    """.format(h_pad=hs_pad, w_pad=ws_pad)
                if ws_pad > 0:
                    c += """
                        LEFT_PAD [b, h, w , c : B, H, WE, C ] = =(I[b, h, {w_pad_l}-w,            c]), w < {w_pad_l} ;
                        HCENTER  [b, h, w , c : B, H, WE, C ] = =(I[b, h, w-{w_pad_l},            c]), w < W+{w_pad_l}-1 ;
                        RIGHT_PAD[b, h, w , c : B, H, WE, C ] = =(I[b, h, 2*W - (w-{w_pad_l}) -2, c]);
                        LCR = LEFT_PAD+HCENTER+RIGHT_PAD;
                    """.format(hs_pad=hs_pad, h_pad_t=h_pad[0], h_pad_d=h_pad[1], ws_pad=ws_pad, w_pad_l=w_pad[0], w_pad_r=w_pad[1])
                else:
                    c += "LCR = I;"
                if hs_pad > 0:
                    c += """
                        TOP_PAD   [b, h, w , c : B, HE, WE, C ] = =(LCR[b, {h_pad_t}-h,            w, c]), h < {h_pad_t};
                        VCENTER   [b, h, w , c : B, HE, WE, C ] = =(LCR[b, h-{h_pad_t},            w, c]), h < H+{h_pad_t}-1 ;
                        BOTTOM_PAD[b, h, w , c : B, HE, WE, C ] = =(LCR[b, 2*H - (h-{h_pad_t}) -2, w, c]);
                        TVB = TOP_PAD+VCENTER+BOTTOM_PAD;
                    """.format(hs_pad=hs_pad, h_pad_t=h_pad[0], h_pad_d=h_pad[1], ws_pad=ws_pad, w_pad_l=w_pad[0], w_pad_r=w_pad[1])
                else:
                    c += "TVB = LCR;"
                c += "O = TVB; }"
                inp_dims = input_.shape.dims
                out_dims = (inp_dims[0], inp_dims[1]+hs_pad, inp_dims[2]+ws_pad, inp_dims[3])
            else:
                raise NotImplementedError()
        else:
            raise NotImplementedError()
        super(ReflectionPadding2D, self).__init__(c, [('I', input_) ],
                [('O', plaidml.tile.Shape(input_.shape.dtype, out_dims ) )])
```

**Comment by iperov** [comment]
@SleepProgger did u test it?
cannot find usage of `w_pad_r`

**Comment by SleepProgger** [comment]
Yes, i did, results matched tf.pad.
w_pad_r and h_pad_d aren't really required as they are part of ws_pad and hs_pad already.

**Comment by iperov** [comment]
something wrong in your code.
`w_pad_r` is not used in string formatting

**Comment by iperov** [comment]
if you are making tf.pad compatible version, then left and right pad should be implemented correctly.
Its mean different left/right pad values for width for example [1,3] should work.

**Comment by iperov** [comment]
`tf.pad(t1, [ [0,0], [h_pad,h_pad], [w_pad_l,w_pad_r], [0,0] ], 'REFLECT')`

**Comment by iperov** [comment]
ah ok , understood

**Comment by iperov** [comment]
[0,2] values also work?

**Comment by SleepProgger** [comment]
Oh, good point. Missed that case. 0 for left or top doesn't work with the current version. Will fix it up tomorrow.

**Comment by brianretford** [comment]
This is great, if you get a PR out today or tomorrow we can get it into 0.6.0

**Comment by iperov** [comment]
names and args should be better organized for pr ?

>  i think a version working for all dimension like tf.pad would be better though.

I dont think so. `tf.pad` contains various paddings, "CONSTANT", "REFLECT", or "SYMMETRIC"
plaidML has no goal to match tensorflow backend, but only keras.
Keras.backend has no padding functions.
Keras has a layer named `keras.layers.ZeroPadding2D(padding=(1, 1), data_format=None)`
so it is logical to create similar op e.g.`ReflectPadding2D`

ReflectPadding2D very useful to exclude gray line at border in autoencoders.

**Comment by SleepProgger** [comment]
Fixed the version with top, down, left, right padding.
I think the conditional for VCENTER and HCENTER is a bit of a hack, but it works for every combination of left,right,top,down padding from 0 to 10.

```
class ReflectionPadding2D(plaidml.tile.Operation):
    def __init__(self, input_, h_pad, w_pad):
        if K.image_data_format() == 'channels_last':
            if input_.shape.ndims == 4:
                H, W = input_.shape.dims[1:3]
                ws_pad = w_pad[0] + w_pad[1]
                hs_pad = h_pad[0] + h_pad[1]
                if (type(H) == int and hs_pad >= H) or \
                   (type(W) == int and ws_pad >= W):
                    raise ValueError("Paddings must be less than dimensions.")
                c = """ function (I[B, H, W, C] ) -> (O) {{
                        WE = W + {ws_pad};
                        HE = H + {hs_pad};
                    """
                if ws_pad > 0:
                    adds = ["HCENTER"]
                    if w_pad[0] > 0:
                        c += "LEFT_PAD [b, h, w , c : B, H, WE, C ] = =(I[b, h, {w_pad_l}-w, c]), w < {w_pad_l} ;"
                        adds.append("LEFT_PAD")
                    if w_pad[1] > 0:
                        c += "RIGHT_PAD[b, h, w , c : B, H, WE, C ] = =(I[b, h, 2*W - (w-{w_pad_l}) -2, c]);"
                        adds.append("RIGHT_PAD")
                    c += "HCENTER  [b, h, w , c : B, H, WE, C ] = =(I[b, h, w-{w_pad_l}, c]), w < W+{w_pad_l}-(({w_pad_r} == 0 ? 0 : 1));"
                    c += "LCR = {};".format(" + ".join(adds))
                else:
                    c += "LCR = I;"
                if hs_pad > 0:
                    adds = ["VCENTER"]
                    if h_pad[0] > 0:
                        c += "TOP_PAD   [b, h, w , c : B, HE, WE, C ] = =(LCR[b, {h_pad_t}-h, w, c]), h < {h_pad_t};"
                        adds.append("TOP_PAD")
                    if h_pad[1] > 0:
                        c += "BOTTOM_PAD[b, h, w , c : B, HE, WE, C ] = =(LCR[b, 2*H - (h-{h_pad_t}) -2, w, c]);"
                        adds.append("BOTTOM_PAD")
                    c += "VCENTER   [b, h, w , c : B, HE, WE, C ] = =(LCR[b, h-{h_pad_t}, w, c]), h < H+{h_pad_t}-(({h_pad_d} == 0 ? 0 : 1));"
                    c += "O = {};\n".format(" + ".join(adds))
                else:
                    c += "O = LCR;\n"
                c += "}}"
                c = c.format(hs_pad=hs_pad, h_pad_t=h_pad[0], h_pad_d=h_pad[1], ws_pad=ws_pad, w_pad_l=w_pad[0], w_pad_r=w_pad[1])
                c = c.replace(";", ";\n") # just to make the code a bit nicer
                inp_dims = input_.shape.dims
                out_dims = (inp_dims[0], inp_dims[1]+hs_pad, inp_dims[2]+ws_pad, inp_dims[3])
            else:
                raise NotImplementedError()
        else:
            raise NotImplementedError()
        super(ReflectionPadding2D, self).__init__(c, [('I', input_) ],
                [('O', plaidml.tile.Shape(input_.shape.dtype, out_dims ) )])
```

> This is great, if you get a PR out today or tomorrow we can get it into 0.6.0

Are you planing to include the version with left=right padding or the one with customizable padding for each side ? I'd prefer the version more similar to tf.pad as that makes porting tf code a bit easier.

**Comment by SleepProgger** [comment]
So i went ahead and implemented a K version of reflected padding with support for any axis.
When padding == [[0,0], [t,d], [l, r], [0,0]] it is compatible with tf.pad(..., mode="REFLECT")
I'd argue that because keras has no ReflectedPadding layer this should be implemented as backend/K function and not as layer.
This way everyone who want to support plaidml as backend can write a simple keras.Layer class supporting TF and plaidml using the appropriate backend functions.

This *could* be written in tile, but to be honest, i don't think it would make much of a difference performance wise and would increase complexity by a fair bit.
I'd be happy to send this as PR if this way is deemed ok.

```
def reflection_padding(inp, paddings):
    paddings = [(x, x) if isinstance(x, int) else x for x in paddings]
    ishape = inp.shape.dims
    ndims = inp.shape.ndims
    if len(ishape) != len(paddings):
        raise ValueError("Padding dims != input dims")
    last = inp
    _all_slice = slice(None, None, None)
    
    def _get_slices(ndims, axis, slice_):
        ret = [_all_slice for _ in range(ndims)]
        ret[axis] = slice_
        return tuple(ret)
    
    for axis, pads in ((i, x) for i, x in enumerate(paddings) if x[0]+x[1] != 0):
        pad_data = []
        if pads[0]:
            pre = last[_get_slices(ndims, axis, slice(pads[0], 0, -1))]
            pad_data.append(pre)
        pad_data.append(last)
        if pads[1]:
            post = last[_get_slices(ndims, axis, slice(-2, -pads[1]-2, -1))]
            pad_data.append(post)
        last = K.concatenate(pad_data, axis)
        ishape = last.shape.dims
    return last
```

**Comment by iperov** [comment]
SleepProgger srsly implement via slice concat?
1) it is 2 easy
2) it is ballooning amount of ops and increase time of analyzing.

**Comment by iperov** [comment]
> i don't think it would make much of a difference performance

it's mean you didnt test the code at all?  who should test the code? Users ? rofl.

**Comment by SleepProgger** [comment]
That code is tested (for all 2d padding combinations from 0-8 with different input sizes to be exact).
I just didn't compare performance between the K function version and the other one.

While it requires more ops to execute it should produce the same amount of tile operations as the tile version.
Feel free to benchmark both versions (which is a bit unfair tbh. as your version doesn't support pre and post padding and your and my changed tile version only support 2d padding) and post the results.
It wouldn't be that hard to convert it to one pure tile operation.

**Comment by iperov** [comment]
I already tested my versions from main post
and reported that slice-concat produces HUGE amount of ops, even only for 2 dims.
I suppose your multidim version will produce much more ops.

Anyway I cannot use slice-concat version in my production. I hope it will never be part of plaidml.

**Comment by SleepProgger** [comment]
The version i posted only create ops for axis AND sides when it is required. So there won't be much op bloat.
I am going to wait for someone more knowledgeable who can tell me how much overhead there is in the same code spread over two Ops vs the same code in one Op.
If there is in deed significant overhead i'd be happy to translate that code to a tile version.

**Comment by brianretford** [comment]
Right now we fuse elementwise ops  (potentially w/ broadcast) onto the end of contractions. Stripe enables fusion for contractions where it makes sense, though we have some work to do to make fusion happen all the times it can. Can you provide a dump of a the Tile code?

**Comment by SleepProgger** [comment]
Is there a way to dump the complete tile code or only per operation ?

**Comment by brianretford** [comment]
Hey all, I think there's room to have multiple implementations if warranted, and lets be kind and respectful to each other. You can dump it by enabling verbose mode (I think) plaidml._set_internal_vlog(True)

For instance, if @iperov's algorithm is faster for the cases it supports, we can add a switch.

**Comment by SleepProgger** [comment]
First file is test with 1 padding on all sides for x and y.
Seconds file is only pads on the right side.
https://gist.github.com/SleepProgger/07ae35a236694b0c7a451bdee44618cd

The correct function to log is: `plaidml._internal_set_vlog(True)`

I think as the 2D version is a subset of the nD version and doesn't add any overhead a nD version makes more sense as a 2D only one.
If the K version should be abandoned in favor of a tile one that ofc. is fine, but the same remains true.

**Comment by iperov** [comment]
cannot understand who need reflection padding for channels or batches :laughing:

**Comment by SleepProgger** [comment]
Might be interesting to support 1d or 3d content or cases where you don't have batches (That could be solved with a reshape before and after though tbh.). Still it doesn't harm to have it supported i think.

**Comment by iperov** [comment]
amount of ops is not a problem for small networks.
But regarding https://github.com/plaidml/plaidml/issues/274, time of ops reanalyzing has a big impact for user experience.

**Comment by SleepProgger** [comment]
This can be closed as it was added in #294 or am i missing something here ?

**Comment by dgkutnic** [comment]
Thanks @SleepProgger for the catch - we must have missed it in our bug scrubs.

**Comment by iperov** [comment]
@sleepprogger  hi , I created my own ML lib from scratch. Try it https://github.com/iperov/litenn
